lib/LaTeXML/Core/Mouth.pm

```diff
@@ -288,49 +284,53 @@ sub readToken {
       $$self{lineno}++;
       $$self{colno} = 0;
       my $line = $self->getNextLine;
+      # For \read, we have to return something for EOL, and handle implicit final newline
+      my $read_mode = (($STATE->lookupValue('PRESERVE_NEWLINES') || 0) > 1);
+      my $eolch     = "\r";
+      if (my $eol = $STATE->lookupDefinition(T_CS('\endlinechar'))) {
+        $eol   = $eol->valueOf()->valueOf;
+        $eolch = (($eol > 0) && ($eol <= 255) ? chr($eol) : undef); }
       if (!defined $line) {    # Exhausted the input.
+        my $eolcc    = ((defined $eolch) && $STATE->lookupCatcode($eolch)) // CC_OTHER;
+        my $eoftoken = $read_mode && (defined $eolch) && !$$self{at_eof} && $$self{source}
+          && ($eolcc == CC_EOL ? T_CS('\par')
+          : Token($eolch, $eolcc));
         $$self{at_eof} = 1;
         $$self{chars}  = [];
         $$self{nchars} = 0;
+        return $eoftoken if $eoftoken;
         return; }
       # Remove trailing spaces from external sources
       if ($$self{source}) { $line =~ s/ *$//s; }
-      # Then append the appropriate \endlinechar, or "\r"
-      if (my $eol = $STATE->lookupDefinition(T_CS('\endlinechar'))) {
-        # \endlinechar<0 or >255 means no character is appended
-        $eol = $eol->valueOf()->valueOf;
-        $line .= chr($eol) if $eol >= 0 and $eol <= 255; }
-      else {
-        $line .= "\r"; }
+      # Then append the appropriate \endlinechar, or "\r";
+      $line .= $eolch if defined $eolch;
+
       $$self{chars}  = splitChars($line);
       $$self{nchars} = scalar(@{ $$self{chars} });
       # In state N, skip spaces
       while (($$self{colno} < $$self{nchars})
         # DIRECT ACCESS to $STATE's catcode table!!!
         && (($$STATE{catcode}{ $$self{chars}[$$self{colno}] }[0] || CC_OTHER) == CC_SPACE)) {
         $$self{colno}++; }
-
+      return T_MARKER('EOL') if $read_mode && ((!defined $eolch) || ($eolch ne "\r"));
```

Comment:
Well, that really is the end of the line, but it returns the end-line Marker *after* it has prepared the next line (if any) for input.

Reply:
Eheh, I still don't get it. When you run this on the very first line, don't you get a marker *before* all the other tokens?

Reply:
@xworld21 if I am getting your question correctly, the answer is to notice when line 315 is actually entered - which is when the condition on line 283 is true:
```
 if ($$self{colno} >= $$self{nchars}) {
```

So this T_MARKER will get created only in the circumstance where one line has been completed, and the next line has been setup, i.e. at the exact time the Mouth transitions to reading the next line. So you're right this will happen *before* all tokens, but those are the *next* tokens, after the line break.

Reply:
Thanks for that! The test is slightly clunkier, but only returns the marker exactly when needed.

Reply:
Ah, now I see what you were trying to do! It all makes sense now.